boot-qemu.py

```diff
@@ -0,0 +1,247 @@
+                )
+
+        # Kernel options
+        if self.cmdline:
+            self._qemu_args += ['-append', ' '.join(self.cmdline)]
+        self._qemu_args += ['-kernel', self.kernel]
+        self._qemu_args += ['-initrd', self._prepare_initrd()]
+
+        # KVM
+        if self.use_kvm:
+            if not self.smp:
+                self.smp = self._get_default_smp_value()
+            self._qemu_args += ['-cpu', self._kvm_cpu, '-enable-kvm']
+
+        # Machine specs
+        self._qemu_args += ['-m', self._ram]
+        if self.smp:
+            self._qemu_args += ['-smp', str(self.smp)]
+
+        # Show information about QEMU
+        utils.green(f"QEMU location: \033[0m{self._qemu_path.parent}")
+        utils.green(f"QEMU version: \033[0m{self._get_qemu_ver_string()}")
+
+        # Pretty print and run QEMU command
+        qemu_cmd = [qemu_path, *self._qemu_args]
+        print(f"$ {' '.join(shlex.quote(str(elem)) for elem in qemu_cmd)}")
+        subprocess.run(qemu_cmd, check=True)
+
+
+class X86QEMURunner(QEMURunner):
+
+    def __init__(self):
+        super().__init__()
+
+        self.cmdline += ['console=ttyS0', 'earlycon=uart8250,io,0x3f8']
+
+        self._default_kernel_path = Path('arch/x86/boot/bzImage')
+        self._initrd_arch = 'x86'
+        self._qemu_arch = 'i386'
+
+    def _can_use_kvm(self):
+        return platform.machine() == 'x86_64' and self._have_dev_kvm_access()
+
+    def run(self):
+        if self.use_kvm:
+            self._qemu_args += ['-d', 'unimp,guest_errors']
+
+        super().run()
+
+
+class X8664QEMURunner(X86QEMURunner):
+
+    def __init__(self):
+        super().__init__()
+
+        self._initrd_arch = self._qemu_arch = 'x86_64'
+
+    def run(self):
+        if not self.use_kvm:
+            self._qemu_args += ['-cpu', 'Nehalem']
+
+        super().run()
+
+
+def parse_arguments():
+    parser = ArgumentParser(description='Boot a Linux kernel in QEMU')
+
+    parser.add_argument(
+        '-a',
+        '--architecture',
+        choices=SUPPORTED_ARCHES,
+        help='The architecture to boot. Possible values are: %(choices)s',
+        metavar='ARCH',
+        required=True)
+    parser.add_argument(
+        '-k',
+        '--kernel-location',
+        required=True,
+        help='Absolute or relative path to kernel image or build folder.')
+    parser.add_argument(
+        '--no-kvm',
+        action='store_true',
+        help='Do not use KVM for accelration even when supported.')
+
+    return parser.parse_args()
+
+
+if __name__ == '__main__':
+    args = parse_arguments()
+
+    arch_to_runner = {
+        'x86': X86QEMURunner,
+        'x86_64': X8664QEMURunner,
+    }
+    runner = arch_to_runner[args.architecture]()
+
+    if not (kernel_location := Path(args.kernel_location).resolve()).exists():
+        raise FileNotFoundError(
+            f"Supplied kernel location ('{kernel_location}') does not exist!")
```

Comment:
`utils.find_first_file`?

Reply:
I guess that would look something like

```
utils.find_first_file(Path('/'), [Path(args.kernel_location).resolve()])
```

or I could refactor it to make the relative root optional then it would just be

```
utils.find_first_file([Path(args.kernel_location).resolve()])
```

I guess the error message might read kind of weird in the current form but that could be adjusted. I do not really have a strong opinion on this, I do not think the way it is written right now is that bad.

Reply:
Mostly if we have code that is searching 1 to many Paths then producing an error message, perhaps we can encapsulate that and reuse that as much as possible.  I don't feel strongly either way, was just curious if this might be a good place for reuse.

Reply:
I think I will leave this as is for now since I think this is slightly clearer. If we start adding more of these checks, I'll refactor `utils.find_first_file()` in the way I described for more use.